wolfssl/error-ssl.h

```diff
@@ -35,7 +35,40 @@
 #endif
 
 enum wolfSSL_ErrorCodes {
-    WOLFSSL_FIRST_E              = -301,
+    WOLFSSL_FATAL_ERROR          =   -1,   /* must be -1 for backward compat. */
+
+    /* negative counterparts to namesake positive constants in ssl.h */
+    WOLFSSL_ERROR_WANT_READ_E    =   -2,
+    WOLFSSL_ERROR_WANT_WRITE_E   =   -3,
+    WOLFSSL_ERROR_WANT_X509_LOOKUP_E = -4,
+    WOLFSSL_ERROR_SYSCALL_E      =   -5,
+    WOLFSSL_ERROR_ZERO_RETURN_E  =   -6,
+    WOLFSSL_ERROR_WANT_CONNECT_E =   -7,
+    WOLFSSL_ERROR_WANT_ACCEPT_E  =   -8,
+
+#if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL) || \
+    defined(HAVE_WEBSERVER) || defined(HAVE_MEMCACHED)
+
+    WOLFSSL_X509_V_ERR_CERT_SIGNATURE_FAILURE_E  = -7, /* note conflict with
```

Comment:
These don't belong here.
They are only meant to be in X509 verification error not returned through normal APIs.

Reply:
The incumbent implementation of `wolfSSL_ERR_reason_error_string()` returns error strings for these values.  That's why the shadow values are now added in `enum wolfSSL_ErrorCodes`.  It's all one namespace (of numbers) from the point of view of `wolfSSL_ERR_reason_error_string()`.

That said, we can move the corresponding cases out of the main `switch()` in `wolfSSL_ERR_reason_error_string()`, enabling us to remove the shadow values in `enum wolfSSL_ErrorCodes` without provoking `-Wswitch-enum`, but then we lose the useful build time check that the cases are covered and self-consistent.

Either way, it's worth noting I didn't bother setting up error traces for `WOLFSSL_X509_V_*`.  The sign-flipped `_E` variants added to `enum wolfSSL_ErrorCodes` aren't actually used by anything directly, though they're implicitly instrumented.

Reply:
wolfSSL_ErrorCodes are returned errors.
WOLFSSL_X509_V_ERR_CERT_SIGNATURE_FAILURE_E is not and should not have a string associated with them when SetErrorString is called.
Don't mix the enums.

Reply:
But it is all going through wolfSSL_ERR_reason_error_string() as you said. *facepalm*
Well at least you'll get an error string out! Just that there is overlap!
These X509_V_ERRs aren't return error codes and shouldn't come out as strings when SetErrorString is called.

wolfSSL_ERR_reason_error_string() is all set error reasons.
OpenSSL does not include the X509_V_* errors in this and neither should we.

Reply:
Ah ha!  So you're thinking we should remove that behavior of `wolfSSL_ERR_reason_error_string()`!  That's where I was hoping this was headed.